https://www.kaggle.com/datasets/hellbuoy/car-price-prediction/data?select=CarPrice_Assignment.csv

In [41]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

In [2]:
df = pd.read_csv("CarPrice_Assignment.csv")

In [3]:
df

car_ID  symboling                   CarName fueltype aspiration  \
0         1          3        alfa-romero giulia      gas        std   
1         2          3       alfa-romero stelvio      gas        std   
2         3          1  alfa-romero Quadrifoglio      gas        std   
3         4          2               audi 100 ls      gas        std   
4         5          2                audi 100ls      gas        std   
..      ...        ...                       ...      ...        ...   
200     201         -1           volvo 145e (sw)      gas        std   
201     202         -1               volvo 144ea      gas      turbo   
202     203         -1               volvo 244dl      gas        std   
203     204         -1                 volvo 246   diesel      turbo   
204     205         -1               volvo 264gl      gas      turbo   

    doornumber      carbody drivewheel enginelocation  wheelbase  ...  \
0          two  convertible        rwd          front       88.6  ...   
1          two  convertible        rwd          front       88.6  ...   
2          two    hatchback        rwd          front       94.5  ...   
3         four        sedan        fwd          front       99.8  ...   
4         four        sedan        4wd          front       99.4  ...   
..         ...          ...        ...            ...        ...  ...   
200       four        sedan        rwd          front      109.1  ...   
201       four        sedan        rwd          front      109.1  ...   
202       four        sedan        rwd          front      109.1  ...   
203       four        sedan        rwd          front      109.1  ...   
204       four        sedan        rwd          front      109.1  ...   

     enginesize  fuelsystem  boreratio  stroke compressionratio horsepower  \
0           130        mpfi       3.47    2.68              9.0        111   
1           130        mpfi       3.47    2.68              9.0        111   
2           152        mpfi       2.68    3.47              9.0        154   
3           109        mpfi       3.19    3.40             10.0        102   
4           136        mpfi       3.19    3.40              8.0        115   
..          ...         ...        ...     ...              ...        ...   
200         141        mpfi       3.78    3.15              9.5        114   
201         141        mpfi       3.78    3.15              8.7        160   
202         173        mpfi       3.58    2.87              8.8        134   
203         145         idi       3.01    3.40             23.0        106   
204         141        mpfi       3.78    3.15              9.5        114   

     peakrpm citympg  highwaympg    price  
0       5000      21          27  13495.0  
1       5000      21          27  16500.0  
2       5000      19          26  16500.0  
3       5500      24          30  13950.0  
4       5500      18          22  17450.0  
..       ...     ...         ...      ...  
200     5400      23          28  16845.0  
201     5300      19          25  19045.0  
202     5500      18          23  21485.0  
203     4800      26          27  22470.0  
204     5400      19          25  22625.0  

[205 rows x 26 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   car_ID            205 non-null    int64  
 1   symboling         205 non-null    int64  
 2   CarName           205 non-null    object 
 3   fueltype          205 non-null    object 
 4   aspiration        205 non-null    object 
 5   doornumber        205 non-null    object 
 6   carbody           205 non-null    object 
 7   drivewheel        205 non-null    object 
 8   enginelocation    205 non-null    object 
 9   wheelbase         205 non-null    float64
 10  carlength         205 non-null    float64
 11  carwidth          205 non-null    float64
 12  carheight         205 non-null    float64
 13  curbweight        205 non-null    int64  
 14  enginetype        205 non-null    object 
 15  cylindernumber    205 non-null    object 
 16  enginesize        205 non-null    int64  
 1

In [11]:
# добавим колонку с названием компании и удалим колонку с названием модели
df.insert(1, 'Company', df['CarName'].apply(lambda x: x.split()[0]))
df.drop(['CarName'], axis=1, inplace=True)

In [14]:
df['Company'].unique()

array(['alfa-romero', 'audi', 'bmw', 'chevrolet', 'dodge', 'honda',
       'isuzu', 'jaguar', 'maxda', 'mazda', 'buick', 'mercury',
       'mitsubishi', 'Nissan', 'nissan', 'peugeot', 'plymouth', 'porsche',
       'porcshce', 'renault', 'saab', 'subaru', 'toyota', 'toyouta',
       'vokswagen', 'volkswagen', 'vw', 'volvo'], dtype=object)

In [24]:
# как видим, в некоторых названиях допущены ошибки
df['Company'] = df['Company'].str.lower()
df['Company'].replace('alfa-romero', 'alfa-romeo', inplace=True)
df['Company'].replace('maxda', 'mazda', inplace=True)
df['Company'].replace('porcshce', 'porsche', inplace=True)
df['Company'].replace('toyouta', 'toyota', inplace=True)
df['Company'].replace('vokswagen', 'volkswagen', inplace=True)
df['Company'].replace('vw', 'volkswagen', inplace=True)
df['Company'].unique()

array(['alfa-romeo', 'audi', 'bmw', 'chevrolet', 'dodge', 'honda',
       'isuzu', 'jaguar', 'mazda', 'buick', 'mercury', 'mitsubishi',
       'nissan', 'peugeot', 'plymouth', 'porsche', 'renault', 'saab',
       'subaru', 'toyota', 'volkswagen', 'volvo'], dtype=object)

In [30]:
# преобразуем все строковые переменные в индикаторные
features = df.columns[df.dtypes==object].tolist()
df2 = pd.get_dummies(df, prefix=features, columns=features, drop_first=True)

In [33]:
df2

car_ID  symboling  wheelbase  carlength  carwidth  carheight  curbweight  \
0         1          3       88.6      168.8      64.1       48.8        2548   
1         2          3       88.6      168.8      64.1       48.8        2548   
2         3          1       94.5      171.2      65.5       52.4        2823   
3         4          2       99.8      176.6      66.2       54.3        2337   
4         5          2       99.4      176.6      66.4       54.3        2824   
..      ...        ...        ...        ...       ...        ...         ...   
200     201         -1      109.1      188.8      68.9       55.5        2952   
201     202         -1      109.1      188.8      68.8       55.5        3049   
202     203         -1      109.1      188.8      68.9       55.5        3012   
203     204         -1      109.1      188.8      68.9       55.5        3217   
204     205         -1      109.1      188.8      68.9       55.5        3062   

     enginesize  boreratio  stroke  ...  cylindernumber_three  \
0           130       3.47    2.68  ...                     0   
1           130       3.47    2.68  ...                     0   
2           152       2.68    3.47  ...                     0   
3           109       3.19    3.40  ...                     0   
4           136       3.19    3.40  ...                     0   
..          ...        ...     ...  ...                   ...   
200         141       3.78    3.15  ...                     0   
201         141       3.78    3.15  ...                     0   
202         173       3.58    2.87  ...                     0   
203         145       3.01    3.40  ...                     0   
204         141       3.78    3.15  ...                     0   

     cylindernumber_twelve  cylindernumber_two  fuelsystem_2bbl  \
0                        0                   0                0   
1                        0                   0                0   
2                        0                   0                0   
3                        0                   0                0   
4                        0                   0                0   
..                     ...                 ...              ...   
200                      0                   0                0   
201                      0                   0                0   
202                      0                   0                0   
203                      0                   0                0   
204                      0                   0                0   

     fuelsystem_4bbl  fuelsystem_idi  fuelsystem_mfi  fuelsystem_mpfi  \
0                  0               0               0                1   
1                  0               0               0                1   
2                  0               0               0                1   
3                  0               0               0                1   
4                  0               0               0                1   
..               ...             ...             ...              ...   
200                0               0               0                1   
201                0               0               0                1   
202                0               0               0                1   
203                0               1               0                0   
204                0               0               0                1   

     fuelsystem_spdi  fuelsystem_spfi  
0                  0                0  
1                  0                0  
2                  0                0  
3                  0                0  
4                  0                0  
..               ...              ...  
200                0                0  
201                0                0  
202                0                0  
203                0                0  
204                0                0  

[205 rows x 66 columns]

In [57]:
X = df2.drop("price", axis = 1)
y = df2['price']

In [58]:
# нормализуем данные
scaler = MinMaxScaler()
X_normalized = scaler.fit_transform(X)
# делим данные на обучающие и тестовые
X_train, X_test, y_train, y_test = train_test_split(X_normalized, y, test_size=0.2, random_state=15)

In [59]:
scale_target = MinMaxScaler()
y_train = pd.DataFrame(y_train)
y_train = scale_target.fit_transform(y_train)
y_test = pd.DataFrame(y_test)
y_test = scale_target.transform(y_test)

In [60]:
# создаем и обучаем модель
model = LinearRegression()
model.fit(X_train, y_train)

# формируем предсказание
y_pred = model.predict(X_test)

# оцениваем модель
mse_lr = mean_squared_error(y_test, y_pred)
r2_lr = r2_score(y_test, y_pred)

print("Mean Squared Error:", mse_lr)
print("R-squared :", r2_lr)

Mean Squared Error: 0.007697396843795614
R-squared : 0.8553401254717473


MSE измеряет среднеквадратическую разницу между фактическими и прогнозируемыми значениями. Довольно низкое значение указывает, что прогнозы модели довольно близки к фактическим значениям

Коэффициент детерминации составляет около 0.855, что говорит о том, что 85.5% дисперсии целевой переменной объясняется моделью